src/Orleans.Transactions/State/TransactionalState.cs

```diff
@@ -0,0 +1,516 @@
+
+            if (this.log.Count > 0)
+            {
+                LogRecord<TState> lastLogRecord = this.log.Values.Last();
+                this.version = lastLogRecord.Version;
+                this.value = lastLogRecord.NewVal;
+            }
+            else
+            {
+                this.version = TransactionalResourceVersion.Create(0, 0);
+                this.value = new TState();
+            }
+        }
+
+        /// <summary>
+        /// Check with the transaction agent and rollback any aborted transaction.
+        /// </summary>
+        private void Restore()
+        {
+            foreach (var transactionId in this.log.Keys)
+            {
+                if (transactionId > this.storage.State.StableVersion && transactionAgent.IsAborted(transactionId))
+                {
+                    Rollback(transactionId);
+                    return;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Write log in the format needed for the persistence framework and copy to the persistent state interface.
+        /// </summary>
+        private void RecordInPersistedLog()
+        {
+            this.storage.State.Logs.Clear();
+            foreach (KeyValuePair<long, LogRecord<TState>> kvp in this.log)
+            {
+                this.storage.State.Logs[kvp.Key] = kvp.Value.NewVal;
+            }
+        }
+
+        /// <summary>
+        /// Read Log from persistent state interface.
+        /// </summary>
+        private void RevertToPersistedLog()
+        {
+            this.log.Clear();
+            foreach (KeyValuePair<long, TState> kvp in this.storage.State.Logs)
+            {
+                this.log[kvp.Key] = new LogRecord<TState>
+                {
+                    NewVal = kvp.Value,
+                    Version = TransactionalResourceVersion.Create(kvp.Key, 1)
+                };
+            }
+        }
+
+        private async Task Persist(long newStableVersion, long newWriteLowerBound)
+        {
+            try
+            {
+                RecordInPersistedLog();
+
+                // update storage state
+                TransactionalStateRecord<TState> storageState = this.storage.State;
+                storageState.Value = this.value;
+                storageState.Version = this.version;
+                storageState.StableVersion = newStableVersion;
+                storageState.WriteLowerBound = newWriteLowerBound;
+
+                await this.storage.WriteStateAsync();
+
+                this.stableVersion = newStableVersion;
+            }
+            catch (Exception)
+            {
+                try
+                {
+                    await this.storage.ReadStateAsync();
+                    DoRecovery();
+                }
+                catch (Exception ex)
+                {
+                    throw new InconsistentStateException($"Transactional state {this.config.StateName} is in an inconsistent state.  Force reload.", ex);
```

Comment:
I'm not confident in the safety here. How are we ensuring that the `storage.State` mutations don't affect subsequent calls to this activation? I'm still reviewing, so maybe it will become clear soon.

Reply:
All modifications of storage.state should be serially executed via storageExecutor.  I welcome double checking of this of course, I am fallible. :/

Reply:
Jason & I had a discussion and we decided that to ensure safety we will wrap all calls to the `AsyncSerialExecutor` so that they ensure that `this.storage.State` is consistent before continuing.

In this context, consistent means that `State` represents data that was successfully persisted.